File: 01/aoc2024d01.cpp
#include <cassert>
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <memory>
#include <optional>
#include <print>
#include <span>
#include <sstream>
#include <string>
#include <string_view>
#include <vector>

using namespace std;

class Scanner
{
  private:
	char const *m_start;
	char const *m_current;

	void skip_whitespace()
	{
		while (true) {
			char c = *m_current;
			switch (c) {
			case ' ':
			case '\r':
			case '\t':
			case '\n':
				++m_current;
				break;
			default:
				return;
			}
		}
	}

  public:
	Scanner(string_view src) : m_start(src.data()), m_current(m_start) {}

	optional<long> scan_number()
	{
		skip_whitespace();
		m_start = m_current;
		if (*m_current == '\0') {
			return {};
		}
		while (*m_current >= '0' && *m_current <= '9') {
			++m_current;
		}
		return strtol(m_start, nullptr, 10);
	}
};

static unique_ptr<string> read_file(string_view filename)
{
	ifstream infile(filename.data());
	if (infile.is_open()) {
		stringstream buf;
		buf << infile.rdbuf();
		return make_unique<string>(buf.str());
	}
	println(stderr, "Failed to read '{}'.", filename);
	exit(1);
}

static long distance(span<long> left, span<long> right)
{
	assert(left.size() == right.size());
	assert(std::is_sorted(left.cbegin(), left.cend()));
	assert(std::is_sorted(right.cbegin(), right.cend()));

	long distance = 0;
	for (size_t i = 0; i < left.size(); ++i) {
		if (left[i] > right[i]) {
			distance += left[i] - right[i];
		} else {
			distance += right[i] - left[i];
		}
	}
	return distance;
}

static long similarity(span<long> left, span<long> right)
{
	assert(left.size() == right.size());
	assert(std::is_sorted(left.cbegin(), left.cend()));
	assert(std::is_sorted(right.cbegin(), right.cend()));

	long similarity = 0;
	for (size_t i = 0; i < left.size(); ++i) {
		similarity +=
			left[i] * std::count(right.cbegin(), right.cend(), left[i]);
	}
	return similarity;
}

int main(int argc, char *argv[])
{
	if (argc != 2) {
		println(stderr, "Usage: {} input", argv[0]);
		exit(1);
	}
	unique_ptr<string> data = read_file(argv[1]);

	Scanner scanner(*data);
	vector<long> left{};
	vector<long> right{};

	while (true) {
		optional<long> a = scanner.scan_number();
		optional<long> b = scanner.scan_number();
		if (!a || !b) {
			break;
		}
		left.push_back(a.value());
		right.push_back(b.value());
	}

	std::sort(left.begin(), left.end());
	std::sort(right.begin(), right.end());

	println("Distance: {}", distance(left, right));
	println("Similarity: {}", similarity(left, right));
}

File: 01/aoc2024d01.py
#!/usr/bin/env python
import os
import re
import sys


def count(seq, item):
    count = 0
    for x in seq:
        if x == item:
            count += 1
    return count


def distance(list_1, list_2):
    sum = 0
    for x, y in zip(list_1, list_2):
        sum += abs(x - y)
    return sum


def similarity(list_1, list_2):
    sum = 0
    for x in list_1:
        sum += x * count(list_2, x)
    return sum


if __name__ == "__main__":
    if len(sys.argv) != 2:
        print(f"Usage: {sys.argv[0]} input", file=sys.stderr)
        exit(1)
    if not os.path.exists(sys.argv[1]):
        print(f"File not found: '{sys.argv[1]}'", file=sys.stderr)
        exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            data = f.read()
            numbers = [int(x) for x in re.split(r"\W+", data) if x != ""]
            left = list(
                sorted([x[1] for x in enumerate(numbers) if x[0] & 1 == 0])
            )
            right = list(
                sorted([x[1] for x in enumerate(numbers) if x[0] & 1 == 1])
            )
            print(f"Distance: {distance(left, right)}")
            print(f"Similarity: {similarity(left, right)}")

    except Exception as e:
        print(f"Could not read '{sys.argv[1]}': {e}.")
